New features : Normalisation functions and enhanced criteria.

gmagannaDevelop opened the pull request:
I think this should be a light review. There are no modifications to the existing functionalities. I just added some functions to the main API, to perform the log normalisation of raw counts. I also added the extra columns needed to sample from the inferred distributions.

@pauleve I am asking you to review this PR because I think this could be a new release `v0.1.3` ? So that you can use your workflow to publish to `PyPI` and `colomoto`.

Let me know what do you think.

gmagannaDevelop committed:
:construction: Working on extra criteria for synthesis

This parameters will be used to sample the inferred distributions in order to simulate sequencing data.

gmagannaDevelop committed:
:sparkles: Improved criteria to model the three possible distributions.

The main class now contains all the elements needed to sample from the inferred distributions.

gmagannaDevelop committed:
:see_no_evil: ignore data files

.

gmagannaDevelop committed:
:sparkles: Added normalisation functions.

These can be used to process raw counts and obtain the log-normalised data that is used by most analyses.

pauleve reviewed (approved):
I would have put normalization.py directly in `profile_bin` directory, but that's just a matter of taste ;-)

pauleve commented:
:+1: 
There is no hurry for the release, we can do it later.

gmagannaDevelop committed:
:sparkles: added relative probabilities for Gaussian mixtures

to allow simulation via numpy.

gmagannaDevelop committed:
:sparkles: added a new method to ProfileBin class

ProfileBin.clear_r_envir allows the user to remove all objects related to the binarisation instance that are found on the main environment of the embedded R process. This might be needed or practical when using multiple instances of the class as part of an analysis pipeline.

gmagannaDevelop committed:
:zap: more explicative backing file names

This will prevent the user from thinking the backing files come from an untrusted source.

gmagannaDevelop committed:
:sparkles: added extra criteria

.

gmagannaDevelop committed:
:sparkles: automatic dependency installation

The ProfileBin class can now automatically detect and try to install missing R dependencies.

gmagannaDevelop committed:
:alembic: added simulation suite draft script

Simulate distributions.

gmagannaDevelop committed:
:see_no_evil: update gitignore

.

gmagannaDevelop committed:
:alembic: add extra params to criteria

local parameters for modelling zero-inf genes. This approach will be re-evaluated.

gmagannaDevelop committed:
:alembic: experimenting with Zero-Inf reclassification

An optional parameter has been added in order to mask zero entries before computing most of relevant parameters. Only the DropOutRate and DenPeak are calculated before dropping all zeros, if the parameter mask_zero_entries is set to True. These changes are consistent accross the R and Python interfaces.

gmagannaDevelop committed:
:sparkles: added optional parameter 'dor_threshold'

this will be further tested, to see if it can explain the two apparent classes of zero-inf genes.

gmagannaDevelop committed:
:boom: Added methods for simulation.

ProfileBin.simulation_fit() allows to recalculate the criteria to allow the simulation of zero-inflated genes.

gmagannaDevelop committed:
:sparkles: added diagnostic plot module

in order to facilitate the analysis of simulated zero-inf genes.

gmagannaDevelop committed:
:zap: created simulation_criteria

This allows doing a full-scale simulation from the estimated parameters.

gmagannaDevelop committed:
:art: Followed pylint suggested improvements

Removed unnecessary else sections on if clauses that returned directly.

gmagannaDevelop committed:
:bulb: Changed default dor_threshold for ProfileBin.simulation_fit()

from 0.75 to 0.95 i.e. back to the default value used by Curie on the original paper. The use of this parameter should be discussed on the next reunion as well as whilst writting the paper.

gmagannaDevelop committed:
:art: Followed pylint recommendations for code style

removed unused typing.Dict import, renamed variable p to pool (multiprocessing.Pool context manager).

gmagannaDevelop committed:
:wastebasket: deprecate exponential simulation function

for zero-inf genes.

gmagannaDevelop committed:
:wastebasket: Deprecate ZeroInf as exponential extra criteria (R side)

commented out part of the criteria that were useless as they were part of the exponential model for ZeroInf genes. This dropped the execution time of compute_criteria by ~18 seconds on Nestorowa dataset.

gmagannaDevelop committed:
:bug: fixed simulation bug

when calling simulation_fit(), there was a R-side exception if there were no Zero-Inflated genes. This is caused because the bigmemory matrix descriptor file needs to contain at least one entry. It was corrected by skipping the R-side recalculation of compute_criteria for Zero-Inflated genes (binarisation criteria are used for simulation).

gmagannaDevelop committed:
:bug: Fixed orphan R process bug

when an exception was raised whilst calling compute_criteria, all the R processes spawned to perform the parallel computation were not killed. Now, on the Python-side, if an RRuntimeError is raised, an explicit call to snow::stopCluster() is performed in order to kill the orphan R processes.

gmagannaDevelop committed:
:recycle: added static method to reduce boilerplate

.

gmagannaDevelop committed:
:see_no_evil: ignore local docs

.

gmagannaDevelop committed:
:truck: rename module in profile_binr.synthesis

simulation -> sampling. That module only samples from the distributions which is not a real simulation. Right now a real simulation method is being defined and will be wrapped in a new simulation module.

gmagannaDevelop committed:
:building_construction: Isolate exception handling on R side

When an exception was raised within the parallel foreach loop in compute_criteria, the child R workers were not terminated. This was solved by making the parallel cluster object a global via the "<<-" operator. Then the cluster was stopped fromthe python side. Using the tryCatch({}) function, the cleanup after exception handling is done completely on the R side without the use of globals.

gmagannaDevelop committed:
:construction: first functional unimodal conditional simulator

It has been superficially tested and looks promising

gmagannaDevelop committed:
:construction: added verbose flag to simulate unimodal gene

.

gmagannaDevelop committed:
:bug: fixed orphan descriptor matrices problem

When an error ocurred in the parallel loop, the tryCatch() system only killed the newly spawned R processes, but the unlink(descriptor_file) part became unreachable. I integrated the unlink() calls to the finally = {} section of tryCatch().

gmagannaDevelop committed:
:boom: First functional biased simulation

With this functional biased simulation, we have done a pre-validation of the results. Some functions using parameters that have been left as the original arbitrary choices should be refactored.

gmagannaDevelop committed:
:bug: removed arbitrary DOR mask

.

gmagannaDevelop committed:
:construction: First python-side parallel binarisation functional

The performance gains are incredible ! Using sequential, python side apply on the dataframe results in 3x speedup ! An using the parallel version results in 10x speedup.

gmagannaDevelop committed:
:construction: developping CLI for reproducible experiments

.

gmagannaDevelop committed:
:construction: Parser advancing, need to add actions

.

gmagannaDevelop committed:
:alien: Change ProfileBin.fit and ProfileBin.simulation fit

return self in order to be able to chain the calls.

gmagannaDevelop committed:
:rotating_light: Fixed pylint and Rlint warnings

.

gmagannaDevelop committed:
:boom: First functional version of CLI

Minimal examples will be added.

gmagannaDevelop committed:
:boom: First fully functional CLI

.

gmagannaDevelop committed:
:see_no_evil: update .gitignore

.

gmagannaDevelop committed:
:bug: Fixed experiment directory creation bug

args.publish_dir.mkdir(parents=True, exist_ok=True) was in the wrong place, so the directory was not created when calling rna2bool with a config file.

gmagannaDevelop committed:
:construction: working on RNG reproductibility

The random number generator in biased_simulation_from_binary_state() is now fully reproducible. That means that given a fixed seed, the results of calling the function will allways be the same. We would like to ensure that across executions all random number generation runs are fully reproducible, given a fixed seed, for all functions.

gmagannaDevelop committed:
:sparkles: Added small utility module

to enable faster debug and development

gmagannaDevelop committed:
:see_no_evil: update .gitignore

.

gmagannaDevelop committed:
:rotating_light: fix linter warnings, enhance docstrings

.

gmagannaDevelop committed:
:construction: working on ordering at training time

There seems to be no significant difference between the time needed to train and return a unordered dataframe and sorting it just after training. For some applications it might be desirable to have this property. Once it's done, I'll merge it into synthesis.

gmagannaDevelop committed:
:fire: removing criteria ordering experiment code

.

gmagannaDevelop committed:
:fire: remove PROFILE_source_dev.R

PROFILE_source_dev.R replaces PROFILE_source.R

gmagannaDevelop committed:
:pencil2: Fixed spelling mistakes

.

gmagannaDevelop committed:
:truck: Rename functions within synthesis/sampling

Synthesis and sampling will be merged into a single module. I will refactor the whole package to simplify it

gmagannaDevelop committed:
:construction: Working on reproducible runs.

modules synthesis.sampling and synthesis.simulation will be merged into a single one.

gmagannaDevelop committed:
:construction: Preparing synthesis.sampling for merge

The sampling functions are now implementing the module-wide unique random number generator by default. This will allow for the global seeding mechanism to function, enabling reproducible results.

gmagannaDevelop committed:
:boom: Migrate synthesis/* to simulation.py

.

gmagannaDevelop committed:
:fire: Remove profile_binr.synthesis

All the modules there contained had already been merged into the profile_binr.simulation module.

gmagannaDevelop committed:
:sparkles: Added stream_helpers submodule

This submodule is to be used to generate trajectories and reconstruct them using STREAM. It might be removed, renamed, or replaced for the second release of profile_binr.

gmagannaDevelop committed:
:sparkles: Added models submodule

The models that will be used for demonstrating the effectiveness of the synthesis method.

gmagannaDevelop committed:
:sparkles: Added extra utility functions

With some minor changes to existing ones. This version facilitates the creation of simpler, cleaner notebooks.

gmagannaDevelop committed:
:pencil2: Fixed critical typo in simulation.

The function profile_binr.simulation.random_nan_binariser contained a critical typo, making it have a tendency to set more values to zero than it was supposed to.

gmagannaDevelop committed:
:zap: Improved profile_binr.utils.stream_helpers.simulate_from_boolean_trajectory

enhanced n_samples_per_state parameter behaviour

gmagannaDevelop committed:
:sparkles: added new model

core regulation model

gmagannaDevelop committed:
:sparkles: added a new timer context manager

useful to benchmark the time required to binarise, simulate, etc.

gmagannaDevelop committed:
:boom: Major changes to profile_binr.wrappers.probinr

These include changes in the binarisation scheme, docstring, retaining discarded genes (or not), etc.
